src/main/java/org/apache/sysds/runtime/instructions/ooc/MatrixMatrixBinaryOOCInstruction.java

```diff
@@ -0,0 +1,257 @@
+import org.apache.sysds.runtime.instructions.InstructionUtils;
+import org.apache.sysds.runtime.instructions.cp.CPOperand;
+import org.apache.sysds.runtime.instructions.spark.data.IndexedMatrixValue;
+import org.apache.sysds.runtime.matrix.data.MatrixBlock;
+import org.apache.sysds.runtime.matrix.data.MatrixIndexes;
+import org.apache.sysds.runtime.matrix.operators.AggregateBinaryOperator;
+import org.apache.sysds.runtime.matrix.operators.AggregateOperator;
+import org.apache.sysds.runtime.matrix.operators.BinaryOperator;
+import org.apache.sysds.runtime.matrix.operators.Operator;
+import org.apache.sysds.runtime.util.CommonThreadPool;
+
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+
+public class MatrixMatrixBinaryOOCInstruction extends ComputationOOCInstruction {
+
+
+	protected MatrixMatrixBinaryOOCInstruction(OOCType type, Operator op, CPOperand in1, CPOperand in2, CPOperand out, String opcode, String istr) {
+		super(type, op, in1, in2, out, opcode, istr);
+	}
+
+	public static MatrixMatrixBinaryOOCInstruction parseInstruction(String str) {
+		String[] parts = InstructionUtils.getInstructionPartsWithValueType(str);
+		InstructionUtils.checkNumFields(parts, 4);
+		String opcode = parts[0];
+		CPOperand in1 = new CPOperand(parts[1]); // the 1st large matrix (streamed)
+		CPOperand in2 = new CPOperand(parts[2]); // the 2nd large matrix (streamed)
+		CPOperand out = new CPOperand(parts[3]);
+
+		AggregateOperator agg = new AggregateOperator(0, Plus.getPlusFnObject());
+		AggregateBinaryOperator ba = new AggregateBinaryOperator(Multiply.getMultiplyFnObject(), agg);
+
+		return new MatrixMatrixBinaryOOCInstruction(OOCType.MAPMM, ba, in1, in2, out, opcode, str);
+	}
+
+	/**
+	 *
+	 * @param ec execution context
+	 */
+	@Override
+	public void processInstruction( ExecutionContext ec ) {
+		// 1. Identify the inputs
+		MatrixObject min1 = ec.getMatrixObject(input1);
+		MatrixObject min2 = ec.getMatrixObject(input2);
+
+
+		LocalTaskQueue<IndexedMatrixValue> qIn1 = min1.getStreamHandle();
+		LocalTaskQueue<IndexedMatrixValue> qIn2 = min2.getStreamHandle();
+		LocalTaskQueue<IndexedMatrixValue> qOut = new LocalTaskQueue<>();
+		BinaryOperator plus = InstructionUtils.parseBinaryOperator(Opcodes.PLUS.toString());
+		ec.getMatrixObject(output).setStreamHandle(qOut);
+
+		// Result matrix rows, cols = rows of A, cols of B
+		long resultRowBlocks = min1.getDataCharacteristics().getNumRowBlocks();
+		System.out.println("resultRowBlocks: " + resultRowBlocks);
+		long resultColBlocks = min2.getDataCharacteristics().getNumColBlocks();
+		System.out.println("resultColBlocks: " + resultColBlocks);
+
+		ExecutorService pool = CommonThreadPool.get();
+		try {
+			// Core logic: background thread
+			pool.submit(() -> {
+				IndexedMatrixValue tmpA = null;
+				IndexedMatrixValue tmpB = null;
```

Comment:
split this code up into individual methods for the different cases.

Reply:
done